Add x-kubernetes-validator to CRD for expression validation.

cici37 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
/kind api-change

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Add x-kubernetes-validator to CRD for expression validation.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduce `x-kubernetes-validator` to CRD for expression validation.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: <https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language>
```

cici37 commented:
/sig api-machinery

cici37 commented:
/test pull-kubernetes-node-e2e-containerd

cici37 commented:
/assign @jpbetz

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -122,6 +122,17 @@ type JSONSchemaProps struct {
+// CELValidationRules is used for expression validation.
+// Rule represents the validation rule which will be evaluated by CEL.
+// Message represents the message displayed when validation failed.
+type CELValidationRules struct {
```
nit: add a link to the cel spec here?

cici37 replied:
Added and updated the pr.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/generated.proto:
```diff
@@ -28,6 +28,15 @@ import "k8s.io/apimachinery/pkg/runtime/schema/generated.proto";
+// Rule represents the validation rule which will be evaluated by CEL.
+// Message represents the message displayed when validation failed.
+message CELValidationRules {
+  optional string rule = 1;
```
Rule should be required?

cici37 replied:
`required` seems not suggested in proto and proto3 dropped required(ref: https://github.com/protocolbuffers/protobuf/issues/2497#issuecomment-267422550)

jpbetz replied:
Gotcha. Should we instead of a validation rule to require the field? I don't think we want CRD authors creating validation rules that don't have a rule.

cici37 replied:
I think this [line](https://github.com/kubernetes/kubernetes/blob/c9b7552ab17d594d92ece089d80f5738672e5280/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go#L258) make sure the empty validator will not get into postProcess.

jpbetz replied:
Can we add a test showing that an attempt to add an empty validator list entry has the right behavior?

jpbetz reviewed (changes requested):
Are there any tests we can extend to include a x-kubernetes-validator case to show that the extension is now allowed.

fedebongio commented:
/triage accepted

cici37 commented:
/test pull-kubernetes-unit

cici37 commented:
/assign @deads2k

alculquicondor reviewed (commented):
You also need changes in https://github.com/kubernetes/kubernetes/blob/9340e314baa0ada21480a572f40316d04afbfcf5/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go to read the feature gate and drop the field if the feature is disabled. And you need to add the feature gate of course 😃

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go:
```diff
@@ -6895,6 +6895,216 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
+			},
+			wantError: false,
+		},
 	}
```
To try to enable API review to progress, we've fleshed out a [test grid](https://docs.google.com/spreadsheets/d/1LLfCT20R4ukpbGsfoGCJgTnb6qWmmYRB1v246vYzDvs/edit#gid=0) of everything we'd like to have tested for alpha.

@liggitt @alculquicondor until we have all the tests in that grid implemented (I hope to get through them in O(days)), would you be willing this approach for API review?
- Review the [test grid](https://docs.google.com/spreadsheets/d/1LLfCT20R4ukpbGsfoGCJgTnb6qWmmYRB1v246vYzDvs/edit#gid=0)  and reach consensus that if we were to fully implement it we would have something that satisfies API review
- Finish the API review process, but don't merge this PR
- Once we've implemented all the tests and the missing functionality (like escaping/intOrString support) we do another review pass to ensure what we implement matches what we had listed in the grid
- We finalize approval and merge the PR

My hope with this approach is that it allows us to do as much API review work as possible up front (and right now as opposed to having to delay).

alculquicondor replied:
All the API changes outside of what's in the kube-openapi's CEL compiler are already reviewed. You can certainly keep making progress, address the comments, and I'll review ASAP.

I wonder if the folder in kube-openapi can be tagged as requiring an API review as well. I guess we could add the same set of api-approvers as owners.

jpbetz replied:
Per discussion with @sttts we've decided to move the CEL code into Kubernetes. It's all in this PR, which is separated into 7 commits to make it easier to review (it looks big but most of it is vendoring and code moves):

https://github.com/kubernetes/kubernetes/pull/106051

I've also expanded the testing to cover the above mentioned test grid:
https://github.com/kubernetes/kubernetes/pull/106051/commits/6f9691f2e5f34d46169af38fab6bbf204c6c6fd4

Hopefully this bundles things up for review better. Everything is either in this PR or in https://github.com/kubernetes/kubernetes/pull/106051

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -161,6 +161,18 @@ type JSONSchemaProps struct {
+	// Rule represents the validation rule which will be evaluated by CEL.
+	Rule string `json:"rule,omitempty" protobuf:"bytes,1,opt,name=rule"`
+	// Message represents the message displayed when validation failed.
+	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
```
The message returned to user would be the detailed error together with the ValidationRule.Message if it's set. I think it should be ok if user does not set message(they will still get error message)

jpbetz replied:
I think optional is appropriate. We expect a significant portion of the rules to be short and self descriptive. For these rules, I don't expect developers to give a message that is more readable than the rule itself, especially when they're only adding a message because the API requires it.

alculquicondor replied:
As long as we include the rule as part of the CRD validation error, I suppose that's fine.

cici37 commented:
/hold
Put the pr on hold until the [test grid](https://docs.google.com/spreadsheets/d/1LLfCT20R4ukpbGsfoGCJgTnb6qWmmYRB1v246vYzDvs/edit#gid=0) be reviewed and completed.

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +229,107 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+			schema.Properties[def] = jsonSchema
+		}
+	}
+	if len(schema.Definitions) != 0 {
```
same here

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +229,107 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+			}
+		}
+	}
+	if schema.Dependencies != nil {
```
same here

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +229,107 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+	}
+	if len(schema.Properties) != 0 && !inUse {
+		for _, jsonSchema := range schema.Properties {
+			if !inUse {
```
what about

```golang
if checkValidations(&jsonSchema) {
  return true
}
```

jpbetz committed:
Pin new dependency: github.com/google/cel-go v0.9.0

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +911,22 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+		} else {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validator"), schema.XValidator, "x-kubernetes-validator cannot be empty"))
+		}
+	}
```
I believe we can enforce #1 when we compile the CEL program by only providing field access to the spec, status, apiVersion, kind and metadata.name fields.I agree that being able to check spec and status fields in a single validation rule seems useful.

sttts replied:
yes, we discussed that elsewhere that we can control the exposed metadata fields. Then @liggitt 's use-case with spec and status is fine.

sttts replied:
To give an example, I don't want spec to depend on finalizers. Or some update depending on managedFields and the actor. We should probably make a list of ObjectMeta fields that are purely api-machinery mechanics and part of the "this is a kube object with uniform behaviour" versus those which are not of that nature.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -123,14 +123,21 @@ type JSONSchemaProps struct {
 	// Rule represents the validation rule which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// The validator will be scoped to the location of the x-kubernetes-validations extension in the schema
+	// and will use `self` to represent the scoped field name.
```
We're adding it for Kubernetes, it's not part of the CEL spec. CEL only requires that you don't use any of the lexer's RESERVED keywords (xref: https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax). We need some way to deal with object field names that collide with these keywords, hence the escaping. Here's the [implementation](https://github.com/kubernetes/kubernetes/pull/106051/commits/a84fd66ff5665486ce94373bd0d630482e8c98d0?authenticity_token=MYvxig2QQU6Uy6u%2BUtre8at%2BuJt3pLZj71A8NYAu58J5VobBKOOvcWNRNumStutDDyrdeXfqm0HwVgtm2sM3uw%3D%3D&file-filters%5B%5D=.go#diff-9666b82ca99585fda3235e21e78c5e6b0423ea389c189f656f05a566a02f7552).

alculquicondor replied:
Then that fact should be part of the documentation.

cici37 replied:
It is part of the KEP. And I added a short comment here as well.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1430,3 +1448,17 @@ func SchemaHasInvalidTypes(s *apiextensions.JSONSchemaProps) bool {
+
+// compileCel is for cel compilation check against x-kubernetes-validator rules
+func compileCel(schemaProps *apiextensions.JSONSchemaProps) error {
+	structural, err := structuralschema.NewStructural(schemaProps)
```
Not implemented yet, but we're moving the CEL code to all use structural schemas and will hooking into the CRD validation [here](https://github.com/kubernetes/kubernetes/blob/ec5919cb679696299a452437fcfc69f9c735d4ef/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L709) so we should have this right shortly.  I'll add a test for this in our integration tests.

cici37 replied:
I believe the structural check for `x-kubernetes-validators` fields is done [here](https://github.com/kubernetes/kubernetes/blob/a0d026a6859130dbc7d8b8d94c0847dca39b4657/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L914)

cici37 commented:
Since we agreed on moving cel dependency and CR validation code from k8s.io/kube-openapi to k/k. I have updated the pr reflect the changes. Changes are:
- Port PRs merged into kube-openapi into k/k instead, and integrate with Structural schema types instead of kube-openapi schema types:
  - Fork in part of github.com/google/cel-policy-templates-go https://github.com/kubernetes/kube-openapi/pull/253 (thirdparty forked code for CEL with OpenAPIv3)
  - https://github.com/kubernetes/kube-openapi/pull/257 (Compilation)
- Add x-kubernetes-validations together with CRD validation and feature gate handling

/unhold

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -123,14 +123,21 @@ type JSONSchemaProps struct {
-// ref: https://github.com/google/cel-spec
-type CELValidationRules struct {
+// ValidationRules describe a list of validation rules for expression validation
+//  +listType=set
```
`map` sgtm if you think it adds value.

cici37 replied:
Updated to use `listMapKey=rule`. Thanks for the input^

cici37 commented:
/test pull-kubernetes-typecheck

cici37 commented:
Updates:
- `x-kubernetes-validations` can not on metadata other than metadata.name. We get it for free when CRD is structural [here](https://github.com/kubernetes/kubernetes/blob/359b722c19ff163fb267f2747d794bd667d4f91d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go#L137)
- Added `x-kubernetes-validations` in `allowedFieldsAtRootSchema`
- CRD structural check is done by [opts.requireStructuralSchema](https://github.com/kubernetes/kubernetes/blob/41ab2ccfb08a34926391f700e55b246714a6507d/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L65)

It is ready for another round of review. Please have a look when you have time @alculquicondor @sttts Thank you :)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -0,0 +1,104 @@
+// ScopedVarName is the variable name assigned to the locally scoped data element of a CEL valid.
+const ScopedVarName = "self"
+
+// Compile compiles all the CEL validation rules in the CelRules and returns a slice containing a compiled program for each provided CelRule, or an array of errors.
```
... compiles all the XValidations rules (without recursing into the schema) ...

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -0,0 +1,104 @@
+	// if the type is object, will traverse each field in the object tree and declare
+	if root.IsObject() {
+		for k, f := range root.Fields {
+			// TODO: There is a much larger set of identifiers that collide if unnested
```
shouldn't we decide for one option as a pre-req of persisting any data into etcd ?

sttts replied:
My taste:
- 3 is ugly
- 2 is a little better, but requires that the reserved key words never change, and we must document which these are.
- 1: which escape mechanism would you use?

alculquicondor replied:
I thought the KEP already proposed to use `_`. Although, I think using `self` is more clear.

cici37 replied:
The escaping implementation would be added in https://github.com/kubernetes/kubernetes/pull/106051

jpbetz replied:
I've lost access to the list of options that kicked off this thread, so I don't know what exact options are being compared, but for escaping, my current thinking is:

- We have to deconflict with "RESERVED" CEL lexer keywords somehow. My proposal is that we prefix with _.  This will be most noticeable with 'namespace', which is both a reserved keyword in CEL and a core concept in kubernetes.  To ensure prefixed identifer doesn't collide with other field names, we prefix anything prefixed with _ with N+1 _s.  This rule applies to all field names used anywhere (since the lexer is aware of grammatical context).

That's the easy part, I think. The harder part is what we do about the property names that we binds as variables into the root expression scope. The options for this are:

1. We stop binding properties in the root scope and instead require all properties be accessed via 'self'.
2. We bind property names IF they don't conflict with any other identifiers (incl type names, function names and macro names). If they conflict, we just don't make them available in the root scope.  (This is what we have in the implementation now)
3. We bind property names and leave it up to the parser to determine if they are accessible. 
4. We bind property names, but if they conflict with any other identifiers, we escape them.

I'm currently favoring #1 right now, but it would require a KEP update. My second choice would be #3, but I want to do some testing to see how the parser would behave.

jpbetz replied:
I did some testing of CEL and found:

- CEL can differentiate between function/macro names (`has`, `size`, ...) and identifiers.
- CEL cannot differentiate between type identifiers (`int`, `string`, ...)

So we only need to deal with escaping for the lexers reserved keywords and the relatively small (and fixed) set of type identifiers. I've updated the escaper one my PR.

alculquicondor replied:
> I'm currently favoring #1 right now, but it would require a KEP update.

Are you proposing that we do that forever, or just for a few releases? If just for a few releases, take in consideration that loosening validation in the API is a convoluted process.

IIUC, option 4 is what the KEP initially proposed, right? What is the problem with it now?

sttts commented on api/api-rules/violation_exceptions.list:
```diff
@@ -419,6 +419,7 @@ API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiexten
 API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,XListType
 API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,XMapType
 API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,XPreserveUnknownFields
+API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,XValidations
```
any idea why they still appear here?

cici37 replied:
It is a names_match violation. Thanks Aldo!

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -122,6 +122,26 @@ type JSONSchemaProps struct {
+	// ref: https://github.com/google/cel-spec
+	// The validator will be scoped to the location of the x-kubernetes-validations extension in the schema
+	// and will use `self` to represent the scoped field name.
+	// If an object property collides with a RESERVED keyword, it will be escaped by prepending a _ prefix.
```
we have to list all reserved keywords and this list must never change in the future.

alculquicondor replied:
Can't it be a link to https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax?

cici37 replied:
Added the link. Not sure if "this list must never change in the future". Could we ask user refer to the doc?

alculquicondor replied:
It would be very hard to change the list in the future. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas

Are all the restricted words part of the CEL documentation or do we have extra words?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -161,6 +161,26 @@ type JSONSchemaProps struct {
+
+	// x-kubernetes-validations describes a list of validation rules for expression validation.
+	// This field is an alpha-level. Using this field requires the feature gate `CustomResourceValidationExpressions` to be enabled.
+	XValidations ValidationRules `json:"x-kubernetes-validations,omitempty" protobuf:"bytes,44,rep,name=xKubernetesValidations"`
```
please sync the godoc with internal types

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -161,6 +161,26 @@ type JSONSchemaProps struct {
+	// and will use `self` to represent the scoped field name.
+	// If an object property collides with a RESERVED keyword, it will be escaped by prepending a _ prefix.
+	Rule string `json:"rule" protobuf:"bytes,1,opt,name=rule"`
+	// Message represents the message displayed when validation failed.
```
can we give an example here of the format expected? Is this like "must be X", e.g. "must be a URL with the host matching spec.host" ?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +910,28 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+	if len(schema.XValidations) > 0 {
+		structural, err := structuralschema.NewStructural(schema)
+		if err != nil {
+			allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
```
This leads to duplicated errors for non-structural schemas.

Do we care? This would mean we have an old CRD here that survived from an update when v1beta1 was still a thing.

cc @liggitt

alculquicondor replied:
Would this actually be an `InternalError` or a user error? If the second, you should use `field.Invalid`.

cici37 replied:
For non-structural schemas, XValidations is blocked from adding.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +910,28 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+		} else {
+			compResults := celcomplilation.Compile(structural)
+			if compResults.Error != nil {
+				allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), compResults.Error))
```
can we pin errors to the rule in question?

sttts replied:
or is that part of the error string?

sttts replied:
we could return field errors instead of normal errors from Compile

sttts replied:
Nvmd. The second case is what I meant.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -19,6 +19,7 @@ package customresourcedefinition
 import (
 	"context"
 	"fmt"
+	"k8s.io/apiserver/pkg/features"
```
nit: order

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +229,104 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+func dropDisabledFields(newCRD *apiextensions.CustomResourceDefinition, oldCRD *apiextensions.CustomResourceDefinition) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CustomResourceValidationExpressions) && !xValidationsInUse(oldCRD) {
+		if newCRD.Spec.Validation != nil && newCRD.Spec.Validation.OpenAPIV3Schema != nil {
+			dropXValidationsField(newCRD.Spec.Validation.OpenAPIV3Schema)
```
we also have the per-version validations.

cici37 replied:
Added pre-version validations

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +229,104 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+	}
+	for def, jsonSchema := range schema.Definitions {
+		dropXValidationsField(&jsonSchema)
+		schema.Definitions[def] = jsonSchema
```
definitions are not used today.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +229,104 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+}
+
+// xValidationsInUse returns true if fields XValidations are set
+func xValidationsInUse(crd *apiextensions.CustomResourceDefinition) bool {
```
we also have per-version validations

sttts commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -42,12 +42,13 @@ func SchemaDeclType(s *schema.Structural) *DeclType {
-			"spec": {Name: "spec", Type: DynType},
+			"metadata": {Name: "metadata", Type: NewObjectType("metadata", map[string]*DeclField{
+				"name": {Name: "name", Type: StringType},
+			})},
```
👍

cici37 replied:
Added generatedName as well

sttts commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -42,12 +42,13 @@ func SchemaDeclType(s *schema.Structural) *DeclType {
+				"name": {Name: "name", Type: StringType},
+			})},
+			"spec":   {Name: "spec", Type: DynType},
 			"status": {Name: "status", Type: DynType, Required: false},
```
hmm, these are strange. Spec and status are not something always existing. And they could even be of different type, e.g. `spec` of string type. Technically possible.

jpbetz replied:
DynType means the contents of the field can be anything and that all type checking is deferred to runtime.

That said, looks like there might be problems with what I've put here. Can there be other fields at the same level of spec? If so, my only option is to make the entire embedded object DynType.  I don't think it's possible to say it's _mostly_ dynamic but that there are specific properties we know exist..

sttts replied:
we have no rule that enforces top-level field names. So this is more like additionalProperties, with exception of kind/apiVersion/metadata.

jpbetz replied:
I don't think I can tell the CEL compiler that I want certain fields on an object to have specific types and that any others should be considered additional properties (@TristonianJones correct me if I'm wrong here).  I can do this though:

- Restrict runtime access to only the metadata name and metadata generateName fields
- Use DynType for the embedded, which still allows `embedded.spec.field` type field selection.

This isn't perfect. A CEL program like `embedded.metadata.labels.exists(...)` would compile (ideally it would fail compilation with a 'undefined field: labels' error), but this would always result in a 'no such key' error for labels at runtime.

jpbetz replied:
I've tried to address this. See https://github.com/kubernetes/kubernetes/pull/105477#discussion_r743054690

TristonianJones replied:
I think I supported a set of structured properties with an `additionalProperties` field could, in theory, refer to `map(string, dyn)` or maybe just `dyn`.  Though maybe I removed support for this since I didn't think this was a good experience. You'd just have to modify the `TypeProvider` to support `additionalProperties` on a well-defined object type, so that the `FindFieldType` call would fallback to whatever type is indicated in `additionalProperties`

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -168,7 +168,7 @@ type JSONSchemaProps struct {
 
 // ValidationRules describe a list of validation rules for expression validation
-//  +listType=atomic
+//  +listMapKey=rule
```
a key for atomic does not make sense. I don't think we have a map here and not a set. Atomic sounds fine.

cici37 replied:
https://github.com/kubernetes/kubernetes/pull/105477#discussion_r740558116
From Joe's comment:
```
I'm a bit torn on this one. On one hand, I don't really see the schema of a CRD as being something that can be sub-divided across multiple owners working independently. On the other hand, we made CEL validation rules a list of rules to allow each rule to be authored independently.

Since it looks like we already allow for non-atomic modification of OpenAPIv3 schemas, I'm slightly in favor of listMapKey=rule, but would like to hear what others think.
```

cici37 replied:
Updated to use 
```
// +listType=map
// +listMapKey=rule
```
for now. Please let me know if atomic is strongly preferred.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -702,7 +702,8 @@ func validateCustomResourceDefinitionValidation(customResourceValidation *apiext
 		}
 
-		allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(schema, fldPath.Child("openAPIV3Schema"), openAPIV3Schema, true, &opts)...)
+		schemaIsNonStructural := schemaIsNonStructural(schema)
```
how does this relate to `requireStructuralSchema` below? Cant we just use `requireStructuralSchema` instead?

cici37 replied:
Updated to use `requireStructuralSchema` to enforce XValidations only in structural schema.
Please let me know if it needs more validation.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -911,20 +912,25 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
-		structural, err := structuralschema.NewStructural(schema)
-		if err != nil {
-			allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
+		if schemaIsNonStructural {
```
We should just forbid `XValidations` for non-structural schemas, not even check them. This case can only happen when the CRD at hand survived from etcd through upgrades, i.e. when v1beta1 CRDs were still a thing. It would simplify this code a lot.

alculquicondor replied:
+1

cici37 replied:
`requireStructuralSchema` is used to ensure XValidations only used in structural schema. `requireStructuralSchema` will be false when oldCRD is non-structural(like v1beta1 CRDs) and we will return error if trying to update with XValidations existing without further checking.

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -36,6 +36,7 @@ import (
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	celcomplilation "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel"
```
is there a conflict?
If not, prefer to just use `cel`

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +910,28 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+			} else {
+				for _, cr := range compResults.Results {
+					for _, err := range cr.Errors {
+						if strings.Contains(err.Error(), "rule is not specified") {
```
Using string comparisons is dangerous, as the message can be changed in the future and people might forget to update this check.

You should be exporting error from the cel package to use `errors.Is` or provide a function that tells you what type of error it is, based on the type or an enum stored in the error.

cici37 replied:
Updated to have error type enum for comparison

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +910,28 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+						if strings.Contains(err.Error(), "rule is not specified") {
+							allErrs = append(allErrs, field.Required(fldPath.Child("x-kubernetes-validations").Index(cr.RuleIndex).Child("rule"), "rule is not specified"))
+						} else {
+							allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations").Index(cr.RuleIndex).Child("rule"), err))
```
`field.Invalid`

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -31,11 +32,25 @@ const ScopedTypeName = "apiextensions.k8s.io.v1alpha1.ValidationExpressionSelf"
 // Compile compiles all the CEL validation rules in the CelRules and returns a slice containing a compiled program for each provided CelRule, or an array of errors.
-func Compile(s *schema.Structural) ([]cel.Program, []error) {
-	var allErrors []error
+func Compile(s *schema.Structural) CompilationResults {
```
what about `([]compilationResult, error)`? Just to avoid structs with similar names.

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -75,30 +90,39 @@ func Compile(s *schema.Structural) ([]cel.Program, []error) {
-			if err != nil {
-				allErrors = append(allErrors, fmt.Errorf("program instantiation failed for valid: %v with message: %v", rule, err))
+			ast, issues := env.Compile(rule.Rule)
+			if issues != nil {
```
we are better off returning a single error with `issues.Err()` It will be more readable.

alculquicondor replied:
WDYT of this?

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -19,6 +19,7 @@ package customresourcedefinition
 import (
 	"context"
 	"fmt"
```
empty line after "fmt"

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -912,25 +911,20 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
-			if err != nil {
-				allErrs = append(allErrs, field.Forbidden(fldPath.Child("x-kubernetes-validations"), "x-kubernetes-validations is only supported in structural schema: "+err.Error()))
+			compResults := cel.Compile(structural)
+			if compResults.Error != nil {
```
Is this actually an internal error or user error?

cici37 replied:
Updated the error as following:
1. XValitions specified with non-structural schema: `Invalid`
2. schema failed to convert to structural(ideally should not happen): `Invalid`
3. compilation setup failure(e.g. cel env initialize failure): `InternalError`
4. Rule is not specified: `Required`
5. Compilation failed for one rule: `Invalid`
Please let me know if anything needs to be changed. Thank you 🙏

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/env.go:
```diff
@@ -0,0 +1,209 @@
+// Copyright 2020 Google LLC
```
I was checking other forked files. They seem not update to use kubernetes copyright. [sample file](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/third_party/forked/golang/json/fields.go)

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -0,0 +1,104 @@
+	var propDecls []*expr.Decl
+	var root *celmodel.DeclType
+	var ok bool
+	env, _ := cel.NewEnv()
```
Added catch

cici37 commented:
Updates after last review:
- Error handling:
  - Have defined type for cel compilation error
  - XValitions specified with non-structural schema: `Invalid`
  - schema failed to convert to structural(ideally should not happen): `Invalid`
  - compilation initialization failure(e.g. cel env initialization failure): `InternalError`
  - Rule is not specified: `Required`
  - Compilation failed for one rule: `Invalid`
- Use `requireStructuralSchema` to check if XValidations is allowed
- Added per-version validations into strategy
- Added eval result check from https://github.com/kubernetes/kubernetes/pull/106051#discussion_r741425907
- Small updates
  - Added preserved keyword ref link
  - Added message example
  - Remove struct CompilationResults to avoid struct with similar names
  - Added metadata.generatedName into cel schema declaration
- Currently we use 
    ```
    // +listType=map
    // +listMapKey=rule
    ```   
    since XValidations is a list of rules to allow each rule to be authored independently. Please let me know if `atomic` is strongly preferred.

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,114 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+	if schema.AdditionalProperties != nil {
+		dropXValidationsField(schema.AdditionalProperties.Schema)
+	}
+	for def, jsonSchema := range schema.Properties {
```
Prefer the variable name `i` for indexes https://github.com/golang/go/wiki/CodeReviewComments#variable-names

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,114 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+		}
+	}
+	if len(schema.Properties) != 0 {
+		for _, jsonSchema := range schema.Properties {
```
let the for handle empty list

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,114 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+		}
+	}
+	if len(schema.Definitions) != 0 {
+		for _, jsonSchema := range schema.Definitions {
```
same here

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,114 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+			return true
+		}
+		if len(schema.Items.JSONSchemas) != 0 {
+			for _, jsonSchema := range schema.Items.JSONSchemas {
```
same here

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +910,32 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	}
 
+	if len(schema.XValidations) > 0 {
+		if !opts.requireStructuralSchema {
```
Here we were trying to add error if the schema is non-structural with XValidations. Hence `requireStructuralSchema` should be false to indicate there is non-structural schema

alculquicondor replied:
ah, right. Although I think we always set the option to true now.

cici37 replied:
requireStructuralSchema could be false if we try to update non-structural CRD(like left over crd from v1beta1 :))

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -909,6 +910,32 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
+			} else {
+				compResults, err := cel.Compile(structural)
+				if err != nil {
+					allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
```
The err here indicate the error happening before trying to compile a rule hence it should be all InternalError.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,114 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+// are not enabled.
+func dropDisabledFields(newCRD *apiextensions.CustomResourceDefinition, oldCRD *apiextensions.CustomResourceDefinition) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CustomResourceValidationExpressions) && !xValidationsInUse(oldCRD) {
+		if newCRD.Spec.Validation != nil && newCRD.Spec.Validation.OpenAPIV3Schema != nil {
```
I think it is needed for some integration test case.

alculquicondor replied:
but `dropXValidationsField` already handles nil

cici37 committed:
move third_party/forked/celopenapi from kube-openapi to kubernetes

cici37 committed:
Auto update/generated files

cici37 committed:
Add x-kubernetes-validations into CRD validation

cici37 committed:
Manual changes in /third_party/forked/celopenapi including schema switch and escape.

cici37 commented:
I have reorganize the commits for easier review.
- commit to pin dependency:  https://github.com/kubernetes/kubernetes/pull/105477/commits/dafb7f59af1acbbe13eaf6e2c5b8c0be6ceb5e6e
- commit to direct move third_party/forked/celopenapi from kube-openapi to kubernetes: https://github.com/kubernetes/kubernetes/pull/105477/commits/1248a64cda72399357ff123972513392f9385d10
- manual changes under third_party/forked/celopenapi: https://github.com/kubernetes/kubernetes/pull/105477/commits/3d8804d6b594381481d24034c5cba44a86ecd15a
- auto update/generated files: https://github.com/kubernetes/kubernetes/pull/105477/commits/a4b111a2cdfffcd8c67532a635725fb2339d885a
- manual changes to add x-kubernetes-validations into CRD validation: https://github.com/kubernetes/kubernetes/pull/105477/commits/8a54f7efdca9ed6516278d20c465b7333de1ec55

cici37 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

sttts commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -15,58 +15,84 @@
+		return NewObjectType("intOrString", map[string]*DeclField{
+			"strVal": {Name: "strVal", Type: StringType},
+			"intVal": {Name: "intVal", Type: IntType},
+		})
```
how does this look like in expression later? Is there some field telling me which of the two is set?

jpbetz replied:
Example usages (from the tests):

```
has(self.strVal) && self.strVal == '100m'
has(self.intVal) && self.intVal == 1
```

Note that developers could either write this as a single compound validation rule (`... && ...`) or as two separate rules, whichever they find more convenient.

sttts replied:
ack, looks good

sttts commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -15,58 +15,84 @@
+		})
+	}
+	if s.XPreserveUnknownFields {
+		return DynType
```
this is not completely correct I think. You can have XPreserveUnknownFields despite another type, especially types like lists or map. I.e. you can partially specify a schema that way, but stop pruning in a subtree of the ast.

jpbetz replied:
We might be up against the limits of CEL here. DynType here effectively defers all type checking to runtime. My preference would that we have CEL perform static type checking on known fields but then allow all other fields through for runtime type checking, but I don't think that's possible (cc @TristonianJones).

Note that for runtime type checking, developers can still use field selection `preserveUnknownFieldsProp.something` and `has` checks, they just won't get a compilation failure if they mistype a field name.

jpbetz replied:
I've been thinking about this one some more. How about:

We don't allow validation rules on the contents of any unknown data.

This would apply to all unknown fields in "preserve unknown fields" "embedded".

This means we only validate fields that we actually have schema information about, which actually seems pretty consistent. For example, if a field is embedded, we don't run  any of its openapi validation rules do we?

This would be easy to implement too

jpbetz replied:
I've implemented this, PTAL.

sttts replied:
am little undecided which was is better. I can see use for checking something on unknown data, also because it might be that OpenAPI cannot express what the preserve-unknown-fields value should be and CEL could. On the other hand, this opens door for "misuse" by defining esoteric types (e.g. everything with a prime number of fields is allowed).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -15,58 +15,84 @@
+				return StringType // OpenAPIv3 byte format represents base64 encoded string
+			case "binary":
 				return BytesType
 			case "google-duration":
```
is google-duration a thing in kube?

jpbetz replied:
Looks like the official list of formats we use is from kube-openapi:
- https://github.com/kubernetes/kube-openapi/blob/1a6458611d189dc17e98a0824dc92536365efedf/pkg/validation/validate/type.go
- https://github.com/kubernetes/kube-openapi/blob/1a6458611d189dc17e98a0824dc92536365efedf/pkg/validation/strfmt/format.go#L105

I'll remove `google-*` and add some test for the date/time formats to make sure they're working as expected.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:
```diff
@@ -78,27 +104,26 @@ func SchemaDeclType(s *spec.Schema) *DeclType {
 
 	// EnvSchema defines the schema for CEL environments referenced within Policy Templates.
-	envSchema *spec.Schema
+	envSchema *schema.Structural
```
what is a policy template?

jpbetz replied:
It's the library we forked to get this CEL/OpenAPI type integration code: https://github.com/google/cel-policy-templates-go

We haven't 100% stripped out everything do don't need.

jpbetz replied:
I've done a pass and removed a lot more code from this directory.

leilajal commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,97 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+		dropXValidationsField(&jsonSchema)
+		schema.Properties[def] = jsonSchema
+	}
+	if schema.Items != nil {
```
just out of curiosity, do we check the cases where the declared schema is not matched by existing persisted CR data?

cici37 replied:
Basically it drops the field if feature gate is disabled when creation and FG is disabled + no persisted data with the field for update.

cici37 replied:
We check the existing persisted CRD in PrepareForUpdate func above. For CR, we have similar process to check existing data while updating.

leilajal commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go:
```diff
@@ -222,3 +228,97 @@ func MatchCustomResourceDefinition(label labels.Selector, field fields.Selector)
+// dropDisabledFields drops disabled fields that are not used if their associated feature gates
+// are not enabled.
+func dropDisabledFields(newCRD *apiextensions.CustomResourceDefinition, oldCRD *apiextensions.CustomResourceDefinition) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CustomResourceValidationExpressions) && !xValidationsInUse(oldCRD) {
```
Do we have a test to check for cases where the featuregate is off but the field is present?

cici37 replied:
I added the test cases in [strategy_test.go](staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go) :)

cici37 committed:
Address comments

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -139,9 +139,11 @@ type ValidationRule struct {
 	// To prevent this from causing a subsequent collision, all properties with a _ prefix will always be prefixed by __ (generally, N+1 the existing number of _s).
-	// e.g. rule against field "_pre": {"rule": "__pre > 0"}
+	// You should escape the property by prepending a _ prefix. e.g. rule against field "_pre": {"rule": "__pre > 0"}
+	// We will exclude self and reserved identifiers from the root bound variables without escaping.
```
It won't be obvious to developers that "reserved identifiers" refers to a different set than "RESERVED keyword" above. 

The reserved identifiers are `"int", "uint", "double", "bool", "string", "bytes", "list", "map", "null_type", "type"` and are defined in https://github.com/google/cel-spec/blob/master/doc/langdef.md#values.

jpbetz replied:
Maybe something like: "If a property name is "self" or collides with a reserved language identifier ..., it is not escaped, but it is excluded from the root bound variables and it can only be accessed via "self.<property name>"...

cici37 replied:
Added more details in API spec.

liggitt replied:
> Maybe something like: "If a property name is "self" or collides with a reserved language identifier ..., it is not escaped, but it is excluded from the root bound variables and it can only be accessed via "self.{property name}"...

I like this... it's way simpler to reason about than _ escaping, in my opinion

cici37 commented:
All commits in this PR are moved to https://github.com/kubernetes/kubernetes/pull/106051 and will be merged there.

liggitt commented:
> All commits in this PR are moved to #106051 and will be merged there.

there's a lot of open discussions in this PR... should we copy comments to the new PR or wrap up the discussions here?

alculquicondor commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -122,6 +122,32 @@ type JSONSchemaProps struct {
+	// To prevent this from causing a subsequent collision, all properties with a _ prefix will always be prefixed by __ (generally, N+1 the existing number of _s).
+	// You should escape the property by prepending a _ prefix. e.g. rule against field "_pre": {"rule": "__pre > 0"}
+	// We will exclude self and reserved identifiers from the root bound variables without escaping.
+	// e.g. For a field named “self”, it can only be accessed via “self.self”
```
Also add documentation for `oldSelf`

cici37 commented:
/close
merged with #106051 and #106234
